ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -333,6 +356,16 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 				}
 			}
 		}
+
+		if rayJobInstance.Spec.BackoffLimit != nil && *rayJobInstance.Spec.BackoffLimit > 0 &&
+			rayJobInstance.Status.JobDeploymentStatus == rayv1.JobDeploymentStatusFailed &&
+			rayJobInstance.Status.JobStatus == rayv1.JobStatusFailed &&
+			rayJobInstance.Status.Failed != nil &&
+			*rayJobInstance.Status.Failed < *rayJobInstance.Spec.BackoffLimit+1 {
+			rayJobInstance.Status.JobDeploymentStatus = rayv1.JobDeploymentStatusRestarting
```

Comment:
Maybe we can transition the status from `JobDeploymentStatusRunning` to `JobDeploymentStatusRestarting` directly. If we update the status here, it may enter the `if rayJobInstance.Spec.ShutdownAfterJobFinishes && len(rayJobInstance.Spec.ClusterSelector) == 0` statement.

Reply:
> Maybe we can transition the status from JobDeploymentStatusRunning to JobDeploymentStatusRestarting directly. 

I think it'll be confusing to skip the JobDeploymentStatusFailed state though. Job going directly from Running -> Restarting could imply the job was restarted for another reason

> If we update the status here, it may enter the if rayJobInstance.Spec.ShutdownAfterJobFinishes && len(rayJobInstance.Spec.ClusterSelector) == 0 statement.

This is a good catch. Should we instead just add a check for `JobDeploymentStatus == JobDeploymentStatusComplete` here?

Reply:
I think `rayv1.JobDeploymentStatusComplete` and `rayv1.JobDeploymentStatusFailed` should be terminal states, meaning that no state transitions from these two states to any other states are allowed.

Reply:
Maybe consider adding a Failing status, so the state transition would be Running -> Failing -> Restarting -> New ... -> Failed or Complete. This is how the Spark operator does it.

Reply:
https://github.com/kubeflow/spark-operator/blob/5ce3dbacff76bba364055b9b786110f1a4ab3174/pkg/controller/sparkapplication/controller.go#L485-L520

Reply:
I renamed Restarting -> Retrying and updating state flow to Running -> Retrying -> New